spec/controllers/users/mfa_selection_controller_spec.rb

```diff
@@ -0,0 +1,156 @@
+require 'rails_helper'
+
+describe Users::MfaSelectionController do
+  before do
+    allow(IdentityConfig.store).to receive(:select_multiple_mfa_options).and_return(true)
+    user = build(:user, :signed_up)
+    stub_sign_in(user)
+  end
+  let(:current_sp) { create(:service_provider) }
+
+  describe 'GET index' do
+    context 'when the user is using one authenticator option' do
+      it 'shows the mfa setup screen' do
+        controller.user_session[:selected_mfa_options] = ['backup_code']
+
+        get :index
+
+        expect(response).to render_template(:index)
+      end
+    end
+  end
+
+  describe 'Submitting the form' do
+    it 'submits the TwoFactorOptionsForm' do
+      user = build(:user)
+      stub_sign_in_before_2fa(user)
+
+      voice_params = {
+        two_factor_options_form: {
+          selection: 'voice',
+        },
+      }
+      params = ActionController::Parameters.new(voice_params)
+      response = FormResponse.new(success: true, errors: {}, extra: { selection: ['voice'] })
+
+      form = instance_double(TwoFactorOptionsForm)
+      allow(TwoFactorOptionsForm).to receive(:new).with(user).and_return(form)
+      expect(form).to receive(:submit).
+        with(params.require(:two_factor_options_form).permit(:selection)).
+        and_return(response)
+      expect(form).to receive(:selection).and_return(['voice'])
+
+      patch :update, params: voice_params
```

Comment:
this test copies a style we've done in the past that I don't like because it stubs everything. IMO our goal should be to check for observable outcomes, such as checking that the correct values were logged. I tried this out locally and it works:

```suggestion
      user = build(:user)
      stub_sign_in_before_2fa(user)
      stub_analytics

      voice_params = {
        two_factor_options_form: {
          selection: 'voice',
        },
      }
      params = ActionController::Parameters.new(voice_params)
      response = FormResponse.new(success: true, errors: {}, extra: { selection: ['voice'] })

      form = instance_double(TwoFactorOptionsForm)
      allow(TwoFactorOptionsForm).to receive(:new).with(user).and_return(form)

      patch :update, params: voice_params

      expect(@analytics).to have_logged_event(Analytics::USER_REGISTRATION_2FA_SETUP, response.to_h)
```